system/security/digisign/digisign.cpp

```diff
@@ -0,0 +1,247 @@
+
+    virtual ~_CDigitalSignatureManager()
+    {
+    }
+
+    bool isDigiSignerConfigured()
+    {
+        return signingConfigured;
+    }
+
+    bool isDigiVerifierConfigured()
+    {
+        return verifyingConfigured;
+    }
+
+    //Create base 64 encoded digital signature of given text string
+    bool digiSign(const char * text, StringBuffer & b64Signature)
+    {
+        if (!signingConfigured)
+            throw MakeStringException(-1, "digiSign:Creating Digital Signatures not configured");
+
+        CriticalBlock b(digiSignCrit);
+
+        if (privateKeyBuff.isEmpty())
+            privateKeyBuff.loadFile(privateKeyFile.str());
+        if (privateKeyBuff.isEmpty())
+            throw MakeStringException(-1, "digiSign:Cannot load private key file");
+
+#ifdef _USE_OPENSSL
+        //create an RSA object from private key
+        BIO * keybio = BIO_new_mem_buf((void*) privateKeyBuff.str(), -1);
+        if (keybio == nullptr)
+        {
+            EVP_THROW("digiSign:BIO_new_mem_buf: %s");
+        }
+
+        RSA * rsa = PEM_read_bio_RSAPrivateKey(keybio, nullptr, nullptr, (void*)passphraseBuff.str());
+        if (rsa == nullptr)
+        {
+            BIO_free_all(keybio);
+            EVP_THROW("digiSign:PEM_read_bio_RSAPrivateKey: %s");
+        }
+        BIO_free_all(keybio);
+
+        //With the RSA object, create the digest and digital signature
+        EVP_MD_CTX * RSASignCtx = EVP_MD_CTX_create(); //allocate, initializes and return a digest context
+        if (nullptr == RSASignCtx)
+        {
+            EVP_THROW("digiSign:EVP_MD_CTX_create returned NULL: %s");
+        }
+        EVP_PKEY * priKey = EVP_PKEY_new();
+        EVP_PKEY_assign_RSA(priKey, rsa);
+
+        //initialize context for SHA-256 hashing function
+        if (EVP_DigestSignInit(RSASignCtx, nullptr, EVP_sha256(), nullptr, priKey) <= 0)
+        {
+            EVP_CLEANUP(priKey, RSASignCtx);
+            EVP_THROW("digiSign:EVP_DigestSignInit: %s");
+        }
+
+        //add string to the context
+        if (EVP_DigestSignUpdate(RSASignCtx, (size_t*)text, strlen(text)) <= 0)
+        {
+            EVP_CLEANUP(priKey, RSASignCtx);
+            EVP_THROW("digiSign:EVP_DigestSignUpdate: %s");
+        }
+
+        //compute length of signature
+        size_t encMsgLen;
+        if (EVP_DigestSignFinal(RSASignCtx, nullptr, &encMsgLen) <= 0)
+        {
+            EVP_CLEANUP(priKey, RSASignCtx);
+            EVP_THROW("digiSign:EVP_DigestSignFinal1: %s");
+        }
+
+        if (encMsgLen == 0)
+        {
+            EVP_CLEANUP(priKey, RSASignCtx);
+            EVP_THROW("digiSign:EVP_DigestSignFinal length returned 0: %s");
+        }
+
+        //compute signature (signed digest)
+        unsigned char * encMsg = (unsigned char*) malloc(encMsgLen);
+        if (encMsg == nullptr)
+        {
+            EVP_CLEANUP(priKey, RSASignCtx);
+            throw MakeStringException(-1, "digiSign:malloc(%ld) returned NULL",encMsgLen);
```

Comment:
minor: using a size_t in printf could cause issues.  Should either cast to unsigned and use %u, or cast to int 64 and use %" I64F "u